### DISCUSSION TITLE
Enabling Consumer Write/Edit Access on Provider's Data Plane

### DISCUSSION THREAD

**Discussion by MemorySounds** [eclipse-edc/Connector#4651] [Q&A] [answered]
Hi EDC Community,

I’m new to EDC and exploring a use case where a consumer participant wishes to add or modify data on a provider’s data plane after a successful contract negotiation. Reading the documentation, I see only mentioned the "PUSH" / "PULL" types but these only talk about data leaving the Provider and transferring over to a Consumer.

More specifically:

Is it feasible to allow a consumer (say, after a successful contract negotiation) to have _write/edit_ permissions on the provider's data?

If so, has anyone implemented this, are there recommended approaches or examples of handling such operations?
Would this require significant modifications to the data plane, and are there any existing extensions or patterns to handle CRUD operations?

I’m guessing that extending the data plane to support write operations and ensuring the control plane manages the contracts to enforce proper authentication and authorization is the main bit of work.

Any advice, pointers to discussions, or similar use cases would be greatly appreciated.

Thanks in advance :)

**Comment by jimmarino** [comment, marked as answer]
Hi,

Data sharing involves the transfer of data from a provider to a consumer. It doesn't mean data is "shared in place" so it does not make any sense for a consumer to modify data on the provider. Data can be transferred (copied) and modified by the consumer.

This is all extensively covered in the documentation.

**Comment by MemorySounds** [comment]
Hi @jimmarino,

Thanks for your quick response - I appreciate it 👍 

I see what you're saying regarding the data being transferred (copied) from Provider->Consumer (unidirectional transfer of data).

I'd appreciate a bit more clarification though.

I came across the section in the [documentation](https://eclipse-edc.github.io/documentation/for-contributors/data-plane/data-plane-signaling/#3-data-plane-public-api) that describes how providers can expose a REST API for “Http-PULL” transfers. It seems like the same approach could, in theory, allow for other RESTful operations (like POST or PATCH). Is there anything in the EDC’s design philosophy or architecture that explicitly discourages this? Or is it just that this kind of use case hasn’t been a focus so far?

Here’s an example of what I’m imagining:

Say there’s an asset described at /assets/1, with metadata like this:

```
{
  "@id": "https://api.example.com/objects/resource1/",
  "creation_date": "2024-01-01T10:00:00Z",
  "update_date": "2024-01-02T12:00:00Z",
  "title": "Sample Resource",
  "description": "This is a sample description for the resource, providing basic metadata.",
  "copyright": "Example Corp",
  "website": "https://example.com/",
  "licences": {
    "@id": "https://api.example.com/objects/resource1/licences/",
    "@type": "ldp:Container",
    "ldp:contains": [
      {
        "@id": "https://api.example.com/objects/licences/1/",
        "creation_date": "2024-01-01T09:00:00Z",
        "update_date": "2024-01-02T11:00:00Z",
        "name": "Sample License",
        "short_desc": "A brief description of the license.",
        "description": "Details about the license go here.",
        "url": "https://example.com/license-info",
        "@type": "example:License"
      }
    ]
  }
}
```

If the consumer wanted to update the description property, they’d typically send a PATCH request to /assets/1. This kind of write operation obviously falls outside the default data-sharing model of transferring data from provider to consumer, but I was wondering:

 - Could the provider’s public API concept be extended to support write operations like this?
- Would doing so require significant changes to the data plane or control plane? Or is the extensibility of the EDC toolbox flexible enough to handle this kind of bidirectional communication?
- Would an alternative approach—where participants act as both providers and consumers—make sense for enabling write operations, or would that introduce challenges like synchronisation or conflicting edits?

I'm trying to understand if this would be feasible, and if not, is it a philosophical constraint or a technical one. 

Thanks again :)

**Reply by jimmarino** [reply]
I think you are confusing several things:

1. Assets (Datasets) are offered through a catalog by the provider. The provider owns the catalog, a consumer should never be allowed to modify it.
2. Data exchange involves the transfer of data, e.g. "by value" not "by reference". Therefore it makes no sense for the consumer to modify it on the provider side. That would also introduce a massive security vulnerability.

So, in a nutshell, what you are proposing is completely at odds with what a dataspace is intended for. Your use case may require different technology.

These concepts are extensively documented, so I'll keep my comments short. The discussion board is intended for topics dealing with the EDC codebase so if you have further questions related to this, you could ask them on the project Discord channel.
